### ISSUE TITLE
create initial flow for bootstrapping a new 3bot

### ISSUE THREAD

**Issue by rkhamis** [threefoldfoundation/3bot#7]
We need to create a chatflow in the bootstrapping 3bot (loaded automatically in the startup of the flist) that allows a new user to deploy a new 3bot.

Some information that will be needed from the user should be asked here, such as:
 - email address
 - maybe captcha? to be discussed

this information will be used to send email of the newly deployed 3bot to complete initialization.

See [spec](https://github.com/threefoldfoundation/3bot/blob/master/specs/register.md) for more info.

**Comment by serboctor** [comment]
Created the basic chatflow. Now working on the post-processing of the data.

**Comment by serboctor** [comment]
Most of the specs are unanswered questions, so I can't really proceed without these questions answered. I have some proposals/discussions for some of those questions, but some other parts are completely foreign to me. Below are some of the proposals/discussions:

**- Captcha:** currently the chat bot doesn't have any functionality to supports captchas. I propose adding a new question type `ask_captcha`. The chatflow will generate a new captcha image and send its bytes instead of the question, the frontend/client will use the bytes to reform the image, and the chatbot will compare the user's input to the string used initially to generate the captcha.

**- Deployment:** where and how are we going to deploy the user 3bot container? Do I use zeroos client to directly deploy the 3bot? and if that's the case, are we going to have a specific farm for these deployments? may be this farm can be an environment variable and the chat flow uses it to query the capacity. Then using the result from the capacity, I pick a node for deployment based on its location (close to user) and available resource. And for the container deployment on the selected node, do I use the zrobot running on this node to make sure that the container creation can be managed properly and to make sure we have self healing and a safety net in case something fails?
Do we need to store this info (user email paired with the node that has his 3bot container) somewhere?

** -Registeration: ** If we can't register the user 3bot from the bootstrap bot because of needing its private key, can the user 3bot register itself after it boots and send the email to the user instead of relying on the bootstrap bot to do that?  This would require running a mail server on each user 3bot, right? I am not sure if this is a horrible overhead or if it is doable or not?
Alternatively, would it be possible to have an actor running on the user 3bot, this actor does the registration and returns the link. The actor would be on the user 3bot so we won't have a private key issue, but the bootstrap 3bot would be responsible for calling this actor and emailing the user the link received from the actor execution. To make sure the actor doesn't get executed by anyone other than bootstrap 3bot, may be the bootstrap 3 bot can set some sort of environment variable on the container while creating the user 3bot, and the actor will use this variable to make sure that the caller is an authorized caller ie: the bootstrap bot.

The url part is not very clear to me, is it supposed to happen as part of a registration after the user 3bot is registered on the blockchain? and if so where are the webgateways it is talking to? Also, these url specs are referenced in the main specs however @xmonader says that KDS already rejected this flow and that it is outdated. https://github.com/threefoldfoundation/3bot/blob/master/specs/urls.md

**Comment by xmonader** [comment]
**registration/rate limiting/captcha**: there's a module for openresty to do the rate limiting https://github.com/openresty/lua-resty-limit-traffic

**urls** 
in the recent [openpublishing doc](https://docs.google.com/document/d/1np2kuTfIFxn1OlRTcsOspkGIYQsSMijDoddZSsuvmM8/edit?ts=5ca8ca8c) sent 
```
Only use openresty(nginx) & JSX DNS server as infrastructure for web publishing.
This will allow super easy hosting of any www & wiki repo on the 3bot (in other words don't use traefik or coredns or etcd).
```
How does that play with [3bot DNS](https://github.com/threefoldtech/home/issues/114) because that reached nowhere.

> The 3bot's private key should be and stay on the 3bot. We can not register the 3bot on the blockchain ourselves since we need the private key to sign it

option 1:
we use multisig transactions for registration 
flow:
- user X wants to deploy 3bot using the bootstrapping service
- collect the user X information (and automatically send their public key)
- we create machine with its own wallet
- we create 3bot transaction between (threefold -whoever will be deploying 3bots-, 3bot (his keys already there, user X)
- send user X qr code to sign 

option 2:
It really sounds like a 3bot is just another wallet on the blockchain working on user's behalf so don't think we need to do anything with private key of the 3bot ourselves.

e.g
user X wants to deploy a 3bot (where a 3bot is a machine capable of executing transactions on the blockchain) so we create another wallet on the 3bot machine and now if the user wants to reserve application Y that costs 100tft then user X sends coins to the 3bot address and 3bot does the reservation and the whole interaction with the 3bot is protected with the totp

option 3:
we allow the user push their keys so the user creates another wallet and share it info to the 3bot  assumption coming from [here](https://github.com/threefoldfoundation/3bot/blob/master/specs/wallet.md) 


> Since the url is not known until the registration is submitted to the chain, how can we send a link?
which url? if the 3bot url isn't it supposed to be doubleName.grid.tf ?

**Comment by serboctor** [comment]
progress tracker: working on the captcha element in the chatbot.

**Comment by zaibon** [comment]
@serboctor
 
> The chatflow will generate a new captcha image and send its bytes instead of the question, the frontend/client will use the bytes to reform the image, and the chatbot will compare the user's input to the string used initially to generate the captcha.

@jdelrue is that ok for you ?

> Do I use zeroos client to directly deploy the 3bot? 

Since 0-robot is not going to be continued, I would go for that solution yes, direct connection to the node.

> and if that's the case, are we going to have a specific farm for these deployments?

No, 3bot can be deployed anywhere. No farm will have the monopoly of the 3bot hosting.
From the spec I see the user needs to enter a location, so we can pick a random farm close to the location specified and pick the least used node in the farm to deploy a 3bot.

> Do we need to store this info (user email paired with the node that has his 3bot container) somewhere?

That sounds like a reasonable things to do indeed.

@serboctor  Regarding the registration you didn't get it right. The full flow is documented there: https://github.com/threefoldfoundation/3bot/blob/master/specs/images/3Bot_creation_flow.png

I don't really understand what you're saying regarding private key. The private key of a 3bot never leave the 3bot itself. Which means during the creation of the 3bot, the key is generated by the 3bot itself.

The email is sent by the bootstrap bot. Since he's the one who know where the 3bot is deployed he's also the one to forward that information to the user.

> The url part is not very clear to me, is it supposed to happen as part of a registration after the user 3bot is registered on the blockchain? and if so where are the webgateways it is talking to? Also, these url specs are referenced in the main specs however @xmonader says that KDS already rejected this flow and that it is outdated. https://github.com/threefoldfoundation/3bot/blob/master/specs/urls.md

This has nothing to do with dns or anything like that. The URL in this case just refrence an HTTP address with a token generated by the 3bot and sent to the user. So the 3bot can authenticate the user the first time it connects to it.

**Comment by serboctor** [comment]
> I don't really understand what you're saying regarding private key. The private key of a 3bot never leave the 3bot itself. Which means during the creation of the 3bot, the key is generated by the 3bot itself.

>The email is sent by the bootstrap bot. Since he's the one who know where the 3bot is deployed he's also the one to forward that information to the user.


Yes, I understand this. And I know the key won't leave the 3bot. And I also saw and understand the flow. However, my comment/suggestion addresses the `open questions` part in the specs.

**Comment by serboctor** [comment]
> registration/rate limiting/captcha: there's a module for openresty to do the rate limiting https://github.com/openresty/lua-resty-limit-traffic

This looks cool! but this limits the rate and doesn't really handle the authenticity of the request. I think the main goal here is to make sure the request is made by a human being not some spammy bot and in this case a captcha makes more sense, and may be later on some other form of authentication as mentioned in the specs?
I added an `ask_captcha` question to the gedis chatbot. https://github.com/threefoldtech/digitalmeX/commit/bc51d29db3626a23caf0a60ada70a961f960f648

As per our call this morning, the double name will be part of the initialization for now and not the registeration.

**Comment by serboctor** [comment]
Worked on picking an appropriate node and deploying the container. I am testing the whole flow so I can move on to flist creation.

**Comment by serboctor** [comment]
Added sal for managing userbot container creation https://github.com/threefoldtech/jumpscaleX/commit/0f789d38b00a306696a95a3a34dd33ff9f185aef

Created an actor to run on the user bot to allow registeration bot to ask about the initialization token https://github.com/threefoldtech/digitalmeX/blob/88dee724a091fcaf7c2a709e741b6056b47d6697/packages/user_bot/actors/userbot.py

The main registeration flow can be found here, it will need another round of testing once initialization flist is ready. https://github.com/threefoldtech/digitalmeX/blob/88dee724a091fcaf7c2a709e741b6056b47d6697/packages/bootstrap_bot/chatflows/bootstrap_bot.py
